// react-imported-component/dist/es5/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var tslib_1 = require("tslib");
var Component_1 = require("./Component");
exports.ComponentLoader = Component_1.ImportedComponent;
var Component_2 = require("./Component");
exports.ImportedComponent = Component_2.ImportedComponent;
var config_1 = require("./config");
exports.setConfiguration = config_1.setConfiguration;
var context_1 = require("./context");
exports.ImportedStream = context_1.ImportedStream;
var helpers_1 = require("./helpers");
exports.remapImports = helpers_1.remapImports;
var HOC_1 = tslib_1.__importStar(require("./HOC"));
exports.imported = HOC_1.default;
exports.lazy = HOC_1.lazy;
var LazyBoundary_1 = tslib_1.__importDefault(require("./LazyBoundary"));
exports.LazyBoundary = LazyBoundary_1.default;
var loadable_1 = require("./loadable");
exports.assignImportedComponents = loadable_1.assignImportedComponents;
exports.whenComponentsReady = loadable_1.done;
exports.dryRender = loadable_1.dryRender;
exports.loadableResource = loadable_1.getLoadable;
var marks_1 = require("./marks");
exports.drainHydrateMarks = marks_1.drainHydrateMarks;
exports.printDrainHydrateMarks = marks_1.printDrainHydrateMarks;
exports.rehydrateMarks = marks_1.rehydrateMarks;
exports.waitForMarks = marks_1.waitForMarks;
var Module_1 = require("./Module");
exports.ImportedModule = Module_1.ImportedModule;
exports.importedModule = Module_1.importedModule;
var loadByChunkName_1 = require("./loadByChunkName");
exports.loadByChunkname = loadByChunkName_1.loadByChunkname;
var useImported_1 = require("./useImported");
exports.useImported = useImported_1.useImported;
exports.useLazy = useImported_1.useLazy;
exports.useLoadable = useImported_1.useLoadable;
var preloaders_1 = require("./preloaders");
exports.addPreloader = preloaders_1.addPreloader;
exports.default = HOC_1.default;
