// react-imported-component/dist/es5/useImported.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var react_1 = require("react");
var config_1 = require("./config");
var context_1 = require("./context");
var detectBackend_1 = require("./detectBackend");
var loadable_1 = require("./loadable");
var marks_1 = require("./marks");
var utils_1 = require("./utils");
function loadLoadable(loadable, callback) {
    var upd = function () { return callback({}); };
    loadable.loadIfNeeded().then(upd, upd);
}
var WEAK_MAP = new WeakMap();
function executeLoadableEventually(loadable, cb) {
    var tracker = {};
    WEAK_MAP.set(loadable, tracker);
    setTimeout(function () {
        if (WEAK_MAP.get(loadable) === tracker) {
            loadable.reload().then(cb);
        }
    }, 16);
}
function useLoadable(loadable, options) {
    if (options === void 0) { options = {}; }
    var UID = react_1.useContext(context_1.streamContext);
    var _a = react_1.useState(function () {
        if (options.import !== false) {
            if (options.track !== false) {
                marks_1.useMark(UID, loadable.mark);
            }
            loadable.loadIfNeeded();
        }
        return {};
    }), forceUpdate = _a[1];
    react_1.useEffect(function () {
        if (options.import !== false) {
            if (options.track !== false) {
                marks_1.useMark(UID, loadable.mark);
            }
            loadLoadable(loadable, forceUpdate);
        }
    }, [loadable, options.import]);
    if (detectBackend_1.isBackend && !loadable_1.isItReady() && loadable.isLoading()) {
        console.error('react-imported-component: trying to render a component which is not ready. You should `await whenComponentsReady()`?');
    }
    var retry = react_1.useCallback(function () {
        if (!loadable) {
            return;
        }
        loadable.reset();
        forceUpdate({});
        loadLoadable(loadable, forceUpdate);
    }, [loadable]);
    if (process.env.NODE_ENV !== 'production') {
        if (detectBackend_1.isBackend) {
            if (!loadable.done) {
                console.error('react-imported-component: using not resolved loadable. You should `await whenComponentsReady()`.');
            }
        }
    }
    return {
        loadable: loadable,
        retry: retry,
        update: forceUpdate,
    };
}
exports.useLoadable = useLoadable;
function useImported(importer, exportPicker, options) {
    if (exportPicker === void 0) { exportPicker = utils_1.es6import; }
    if (options === void 0) { options = {}; }
    var topLoadable = react_1.useState(loadable_1.getLoadable(importer))[0];
    var postEffectRef = react_1.useRef(false);
    var _a = useLoadable(topLoadable, options), loadable = _a.loadable, retry = _a.retry, update = _a.update;
    react_1.useEffect(function () {
        if (postEffectRef.current) {
            executeLoadableEventually(loadable, function () { return config_1.settings.updateOnReload && update({}); });
        }
        postEffectRef.current = true;
    }, ['hot']);
    if (loadable.error) {
        return {
            error: loadable.error,
            loadable: loadable,
            retry: retry,
        };
    }
    if (loadable.done) {
        return {
            imported: exportPicker(loadable.payload),
            loadable: loadable,
            retry: retry,
        };
    }
    return {
        loading: loadable.isLoading(),
        loadable: loadable,
        retry: retry,
    };
}
exports.useImported = useImported;
function useLazy(importer) {
    var _a = react_1.useState(function () {
        var resolve;
        var reject;
        var promise = new Promise(function (rs, rej) {
            resolve = rs;
            reject = rej;
        });
        return {
            resolve: resolve,
            reject: reject,
            lazyComponent: react_1.lazy(function () { return promise; }),
        };
    })[0], resolve = _a.resolve, reject = _a.reject, lazyComponent = _a.lazyComponent;
    var _b = useImported(importer), error = _b.error, imported = _b.imported;
    react_1.useEffect(function () {
        if (error) {
            reject(error);
        }
        if (imported) {
            resolve(imported);
        }
    }, [error, imported]);
    return lazyComponent;
}
exports.useLazy = useLazy;
